### PR TITLE
RowBatcher - First Draft.

### PR DESCRIPTION
**Pull request by anu3990**
So we can incorporate your pull request, please share the following:
* What issue are you addressing with this pull request?
* Are you modifying the correct branch? (See CONTRIBUTING.md)
* Have you run unit tests? (See CONTRIBUTING.md)
* Version of MarkLogic Java Client API (see Readme.txt)
* Version of MarkLogic Server (see admin gui on port 8001)
* Java version (`java -version`)
* OS and version
* What Changed: What happened before this change? What happens without this change?

### CODE REVIEW

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/RowBatcherImpl.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    private  String jobName;
+    private JobTicket jobTicket;
+    private Calendar jobStartTime;
+    private Calendar jobEndTime;
+    private PlanBuilder.PreparePlan preparePlan;
+    private DatabaseClient client;
+    private AtomicLong lowerBound = new AtomicLong(0);
+    private long maxBatches = Long.MAX_VALUE;
+    private long rowCount;
+    private Long bucketSize;
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/RowBatcherImpl.java]: I'm wondering whether the bucket size can be an integer -- that is, whether we would ever retrieve more rows in a single request than Integer.MAX_VALUE.

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/RowBatcherImpl.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    private String jobId;
+    private  String jobName;
+    private JobTicket jobTicket;
+    private Calendar jobStartTime;
+    private Calendar jobEndTime;
+    private PlanBuilder.PreparePlan preparePlan;
+    private DatabaseClient client;
+    private AtomicLong lowerBound = new AtomicLong(0);
+    private long maxBatches = Long.MAX_VALUE;
+    private long rowCount;
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/RowBatcherImpl.java]: Should we initialize this variable to 0?

Should we expose a getRowEstimate() in the RowBatcher interface that returns this value? The application might make some decisions based on this value.

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/RowBatcherImpl.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    public boolean isStarted() {
+        return started.get();
+    }
+
+    private void getRowCount(){
+
+        JsonNode jn = ((DatabaseClientImpl) client).getServices()
+                .getResource(null, "internal/viewinfo", null, params,
+                        new JacksonHandle())
+                .get();
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/RowBatcherImpl.java]: The value of rowCount used in the calculation on current line 248 is obtained by the internal/viewinfo request.

This method should get the value of the tableID and rowCount properties from the JsonNode and assign them to private fields of the RowBatcherImpl.

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/RowBatcherImpl.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            logger.warn("startJob called more than once");
+            return;
+        }
+        if ( getBatchSize() <= 0 ) {
+            withBatchSize(1);
+            logger.warn("batchSize should be 1 or greater--setting batchSize to 1");
+        }
+        jobTicket = ticket;
+
+        getRowCount();
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/RowBatcherImpl.java]: Once we know the batch size, we can spin up the threads and submit tasks.

As sketched in the technical story, each task should 

1. get the next lower bound
2. calculate the upper bound
3. bind the parameters on the plan to the lower and upper bound values
4. use the RowManager to get the result rows
5. send the result rows to the success listener

This code will be similar to startQuerying() and QueryTask in QueryBatcherImpl

**Comment by anu3990** [comment]
Closing this request for now. Will be opening a new one with unit tests.

**Pull request closed without merging**